source/extensions/filters/http/mcp_router/mcp_router.h

```diff
@@ -0,0 +1,281 @@
+struct BackendResponse {
+  std::string backend_name;
+  bool success{false};
+  uint64_t status_code{0};
+  std::string body;
+  std::string session_id;
+  std::string error;
+};
+
+using AggregationCallback = std::function<void(std::vector<BackendResponse>)>;
+
+/**
+ * Callbacks for handling async HTTP responses from backend MCP servers.
+ * Accumulates response data and invokes completion callback when stream ends.
+ */
+class BackendStreamCallbacks : public Http::AsyncClient::StreamCallbacks,
+                               public std::enable_shared_from_this<BackendStreamCallbacks>,
+                               public Logger::Loggable<Logger::Id::filter> {
+public:
+  BackendStreamCallbacks(const std::string& backend_name,
+                         std::function<void(BackendResponse)> on_complete);
+
+  // AsyncClient::StreamCallbacks
+  void onHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override;
+  void onData(Buffer::Instance& data, bool end_stream) override;
+  void onTrailers(Http::ResponseTrailerMapPtr&& trailers) override;
+  void onComplete() override;
+  void onReset() override;
+
+private:
+  void complete();
+
+  std::string backend_name_;
+  std::function<void(BackendResponse)> on_complete_;
+  BackendResponse response_;
+  bool completed_{false};
+};
+
+class ThreadLocalFanoutManager;
+
+/**
+ * Resources for fire-and-forget fanout that need to outlive the filter.
+ * Stored in ThreadLocalFanoutManager and cleaned up when all backends respond.
+ */
+struct DetachedFanout : public Logger::Loggable<Logger::Id::filter> {
+  DetachedFanout(ThreadLocalFanoutManager& manager) : manager_(manager) {}
+
+  std::shared_ptr<Http::MuxDemux> muxdemux;
+  std::unique_ptr<Http::MultiStream> multistream;
+  std::vector<std::shared_ptr<BackendStreamCallbacks>> callbacks;
+  std::vector<Http::RequestHeaderMapPtr> headers;
+
+  ThreadLocalFanoutManager& manager_;
+};
+
+/**
+ * Thread-local storage for detached fanout operations.
+ * Allows the fire-and-forget fanout to complete after their filter is destroyed.
+ */
+class ThreadLocalFanoutManager : public ThreadLocal::ThreadLocalObject,
+                                 public Logger::Loggable<Logger::Id::filter> {
+public:
+  // Store a detached fanout and return raw pointer for tracking
+  DetachedFanout* store(std::unique_ptr<DetachedFanout> fanout) {
+    DetachedFanout* raw = fanout.get();
+    fanouts_[raw] = std::move(fanout);
+    ENVOY_LOG(debug, "Stored detached fanout, total: {}", fanouts_.size());
+    return raw;
+  }
+
+  // Remove a detached fanout (called when all backends complete)
+  void remove(DetachedFanout* fanout) {
+    fanouts_.erase(fanout);
+    ENVOY_LOG(debug, "Removed detached fanout, remaining: {}", fanouts_.size());
+  }
+
+private:
+  absl::flat_hash_map<DetachedFanout*, std::unique_ptr<DetachedFanout>> fanouts_;
+};
+
+/** Configuration for the MCP router filter, containing backend server definitions. */
+class McpRouterConfig {
```

Comment:
Please move into config.h/cc

Reply:
emm, the config.h/cc are mainly for factory not the filter config, and moving there could create a circular dependency, and most of filters are in the same file, e.g., the router and many, so I will keep here, if it blows up can create a `filter_config.h/c` file.